net/net.go

```diff
@@ -16,7 +16,128 @@ limitations under the License.
 
 package net
 
-import "net"
+import (
+	"fmt"
+	"net"
+)
+
+// ParseCIDRs parses a list of cidrs and return error if any is in valid.
+// order is maintained
+func ParseCIDRs(cidrsString []string) ([]*net.IPNet, error) {
+	cidrs := make([]*net.IPNet, 0, len(cidrsString))
+	for _, cidrString := range cidrsString {
+		_, cidr, err := net.ParseCIDR(cidrString)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse cidr value:%q with error:%v", cidrString, err)
+		}
+		cidrs = append(cidrs, cidr)
+	}
+	return cidrs, nil
+}
+
+// IsDualStackIPs returns if a slice of ips is:
+// - all are valid ips
+// - at least one ip from each family (v4 or v6)
+// - an error is returned if length is less than 2
+// - an error is returned if duplicate ips are found
+func IsDualStackIPs(ips []net.IP) (bool, error) {
+	if len(ips) < 2 {
+		return false, fmt.Errorf("ips length is less than 2")
+	}
+
+	v4Found := false
+	v6Found := false
+	seen := make(map[string]int)
+	for _, ip := range ips {
+		if ip == nil {
+			return false, fmt.Errorf("ip %v is invalid", ip)
+		}
+
+		if _, ok := seen[ip.String()]; ok {
+			return false, fmt.Errorf("ip %v is duplicate", ip)
+		}
+
+		seen[ip.String()] = 0
+
+		if v4Found && v6Found {
+			continue
+		}
+
+		if IsIPv6(ip) {
+			v6Found = true
+			continue
+		}
+
+		v4Found = true
+	}
+
+	return (v4Found && v6Found), nil
+}
+
+// IsDualStackIPsStrings returns if
+// - all are valid ips
+// - at least one ip from each family (v4 or v6)
+// - an error is returned if length is less than 2
+// - an error is returned if duplicate ips are found
+func IsDualStackIPsStrings(ips []string) (bool, error) {
```

Comment:
Do we need the full set of {IP, CIDR} x {<t>, <t>String}  ?  Can we force callers to canonicalize?

Reply:
We don't need the full set, forcing the caller to canonicalize would add more pre-call code that the func does really need to operate correctly.

Reply:
I don't know the callsites, what I means here is that if they all eventually convert `[]string` to `[]net.IPNet` eventually anyway, maybe they should convert first and then call `IsDualStackIPs` and not have a `IsDualStackIPStrings` at all.

Reply:
the assumption that caller will always eventually work with `[]net.IP` or `[]net.IPNet` in case of cidrs. That assumption is arguable does not hold true for all. case in point api validation. an api-server does not really need to convert `cidrs` or `ip` to their true form, an api server just needs to ensure that they are dual stack. What i mean to say some callers will *only* work string form, others maybe converting..

Reply:
OK.  I;d prefer less public functions, but this is not a killer